### PR TITLE
Hide unnecessary error checking from the user

### PR DESCRIPTION
**Pull request by tbu-**
This affects the `set_non_blocking` function which cannot fail for Unix or
Windows, given correct parameters. Additionally, the short UDP write error case
has been removed as there is no such thing as "short UDP writes", instead, the
operating system will error out if the application tries to send a packet
larger than the MTU of the network path.

### CODE REVIEW

**Comment by alexcrichton** [comment]
What's the motivation for doing this? It seems backwards to explicitly remove this error handling.

**Comment by tbu-** [comment]
The motivation for this is that the only errors `ioctl` or `ioctlsocket` can return are related to misuse of the API (passing in pointer to invalid memory, not passing in a correct fd, ...), and bubbling those up to the user is not useful.

There is also no such thing as a partially sent UDP packet, so if the OS returns that it has been partially sent, something has already gone horribly wrong (see also the commit message).

File: src/libstd/sys/windows/mod.rs [hunk 1]
```diff
@@ -192,13 +192,13 @@ pub fn wouldblock() -> bool {
...
-pub fn set_nonblocking(fd: sock_t, nb: bool) -> IoResult<()> {
+pub fn set_nonblocking(fd: sock_t, nb: bool) {
     let mut set = nb as libc::c_ulong;
-    if unsafe { c::ioctlsocket(fd, c::FIONBIO, &mut set) != 0 } {
+    (if unsafe { c::ioctlsocket(fd, c::FIONBIO, &mut set) } != 0 {
         Err(last_error())
     } else {
         Ok(())
-    }
+    }).unwrap();
```
Comment (oli-obk) [inline src/libstd/sys/windows/mod.rs]: shouldn't this panic in the true-path and leave out the else path instead of unwrapping a fresh Result?

Comment (tbu-) [inline src/libstd/sys/windows/mod.rs]: This should be equivalent and this will provide the contents of a potential error.

Comment (oli-obk) [inline src/libstd/sys/windows/mod.rs]: dropping the else-path and calling `unwrap()` on `Err(last_error())` would increase readability imho

Comment (tbu-) [inline src/libstd/sys/windows/mod.rs]: Fixed.

File: src/libstd/sys/windows/mod.rs [hunk 2]
```diff
@@ -192,13 +192,14 @@ pub fn wouldblock() -> bool {
...
-    if unsafe { c::ioctlsocket(fd, c::FIONBIO, &mut set) != 0 } {
-        Err(last_error())
-    } else {
-        Ok(())
+    if unsafe { c::ioctlsocket(fd, c::FIONBIO, &mut set) } != 0 {
+        // The above function should not return an error unless we passed it
+        // invalid parameters. Panic on errors.
+        Err(last_error()).unwrap();
     }
+    Ok(())
```
Comment (oli-obk) [inline src/libstd/sys/windows/mod.rs]: that compiles? the function isn't returning anything anymore. If that compiles, rust has a bug.

Comment (tbu-) [inline src/libstd/sys/windows/mod.rs]: Can't check if it compiles, I don't have a windows machine around. Fixed in a commit amend.

**Comment by Manishearth** [comment]
Needs https://github.com/Manishearth/rust/commit/e61a790495f49580fc0ee831bb07e00b4750b04a to compile on windows.

**Pull request merged**